From b449918e3c1c42f67b2e1e6ec918747398da97df Mon Sep 17 00:00:00 2001
From: Meriipu <Meriipu@users.noreply.github.com>
Date: Sat, 12 Sep 2020 05:27:56 +0200
Subject: [PATCH 4/8] revert the removal of the dropmarker et al.

as of 78 this alone is not enough to have the dropmarker be functional
some additional patches to fix autoOpen on focus/select as well as
avoiding one of the providers is necessary.
---
 browser/base/content/browser.css              |  1 +
 browser/base/content/browser.xhtml            |  3 +++
 browser/components/urlbar/UrlbarInput.jsm     | 19 +++++++++++++++
 browser/components/urlbar/UrlbarView.jsm      |  2 ++
 .../locales/en-US/chrome/browser/browser.dtd  |  2 ++
 .../themes/shared/urlbar-searchbar.inc.css    | 24 +++++++++++++++++++
 6 files changed, 51 insertions(+)

diff --git a/browser/base/content/browser.css b/browser/base/content/browser.css
index 5a00107d8f88..681524a9bd8f 100644
--- a/browser/base/content/browser.css
+++ b/browser/base/content/browser.css
@@ -707,6 +707,7 @@ toolbar:not(#TabsToolbar) > #personal-bookmarks {
 #urlbar-input-container[pageproxystate=invalid] > #page-action-buttons > .urlbar-page-action,
 #identity-box.chromeUI ~ #page-action-buttons > .urlbar-page-action:not(#star-button-box),
 #urlbar[usertyping] > #urlbar-input-container > #page-action-buttons > #urlbar-zoom-button,
+#urlbar[usertyping] > #urlbar-input-container > .urlbar-history-dropmarker,
 #urlbar:not([usertyping]) > #urlbar-input-container > #urlbar-go-button,
 #urlbar:not([focused]) > #urlbar-input-container > #urlbar-go-button {
   display: none;
diff --git a/browser/base/content/browser.xhtml b/browser/base/content/browser.xhtml
index 93b1b0932bf9..eb965d1c58be 100644
--- a/browser/base/content/browser.xhtml
+++ b/browser/base/content/browser.xhtml
@@ -1639,6 +1639,9 @@
                      class="urlbar-icon"
                      onclick="gURLBar.handleCommand(event);"
                      data-l10n-id="urlbar-go-button"/>
+              <image class="urlbar-history-dropmarker urlbar-icon chromeclass-toolbar-additional"
+                     hidden="false"
+                     tooltiptext="&urlbar.openHistoryPopup.tooltip;"/>
               <hbox id="page-action-buttons" context="pageActionContextMenu">
                 <toolbartabstop/>
                 <hbox id="contextual-feature-recommendation" role="button" hidden="true">
diff --git a/browser/components/urlbar/UrlbarInput.jsm b/browser/components/urlbar/UrlbarInput.jsm
index 240dba7461d9..e9f8f182e707 100644
--- a/browser/components/urlbar/UrlbarInput.jsm
+++ b/browser/components/urlbar/UrlbarInput.jsm
@@ -152,6 +152,8 @@ class UrlbarInput {
     }
 
     this.inputField = this.querySelector("#urlbar-input");
+    this.dropmarker = this.querySelector(".urlbar-history-dropmarker");
+    this.dropmarker.hidden = false;
     this._inputContainer = this.querySelector("#urlbar-input-container");
     this._identityBox = this.querySelector("#identity-box");
     this._searchModeIndicator = this.querySelector(
@@ -206,6 +208,7 @@ class UrlbarInput {
       this.addEventListener(name, this);
     }
 
+    this.dropmarker.addEventListener("mousedown", this);
     this.window.addEventListener("mousedown", this);
     if (AppConstants.platform == "win") {
       this.window.addEventListener("draggableregionleftmousedown", this);
@@ -2254,6 +2257,22 @@ class UrlbarInput {
           this.view.autoOpen({ event });
         }
         break;
+      case this.dropmarker:
+        if (event.button != 0) {
+          break;
+        }
+
+        if (this.view.isOpen) {
+          this.view.close();
+        } else {
+          this.focus();
+          this.startQuery({
+            allowAutofill: false,
+            event,
+          });
+          this._maybeSelectAll();
+        }
+        break;
       case this.window:
         if (this._mousedownOnUrlbarDescendant) {
           this._mousedownOnUrlbarDescendant = false;
diff --git a/browser/components/urlbar/UrlbarView.jsm b/browser/components/urlbar/UrlbarView.jsm
index 3448feb45007..25faffab09d8 100644
--- a/browser/components/urlbar/UrlbarView.jsm
+++ b/browser/components/urlbar/UrlbarView.jsm
@@ -384,6 +384,7 @@ class UrlbarView {
 
     this.removeAccessibleFocus();
     this.input.inputField.setAttribute("aria-expanded", "false");
+    this.input.dropmarker.removeAttribute("open");
 
     this.input.removeAttribute("open");
     this.input.endLayoutExtend();
@@ -783,6 +784,7 @@ class UrlbarView {
     this._enableOrDisableRowWrap();
 
     this.input.inputField.setAttribute("aria-expanded", "true");
+    this.input.dropmarker.setAttribute("open", "true");
 
     this.input.setAttribute("open", "true");
     this.input.startLayoutExtend();
diff --git a/browser/locales/en-US/chrome/browser/browser.dtd b/browser/locales/en-US/chrome/browser/browser.dtd
index f6120ac9d0a1..28772c3faf70 100644
--- a/browser/locales/en-US/chrome/browser/browser.dtd
+++ b/browser/locales/en-US/chrome/browser/browser.dtd
@@ -96,6 +96,8 @@ this container is a toolbar. This avoids double-speaking. -->
 <!ENTITY printButton.label            "Print">
 <!ENTITY printButton.tooltip          "Print this page">
 
+<!ENTITY urlbar.openHistoryPopup.tooltip                "Show history">
+
 <!ENTITY searchItem.title             "Search">
 
 <!-- Toolbar items -->
diff --git a/browser/themes/shared/urlbar-searchbar.inc.css b/browser/themes/shared/urlbar-searchbar.inc.css
index a5da19eb186f..1e81aab5fdfa 100644
--- a/browser/themes/shared/urlbar-searchbar.inc.css
+++ b/browser/themes/shared/urlbar-searchbar.inc.css
@@ -530,6 +530,19 @@
   transform: scaleX(-1);
 }
 
+.urlbar-history-dropmarker {
+  list-style-image: url(chrome://global/skin/icons/arrow-dropdown-16.svg);
+  transition: opacity 0.15s ease;
+}
+
+#urlbar[switchingtabs] > #urlbar-input-container > .urlbar-history-dropmarker {
+  transition: none;
+}
+
+#nav-bar:not([customizing="true"]) > #nav-bar-customization-target > #urlbar-container:not(:hover) > #urlbar:not([focused]) > #urlbar-input-container > .urlbar-history-dropmarker {
+  opacity: 0;
+}
+
 #pageActionButton,
 .share-more-button {
   list-style-image: url("chrome://global/skin/icons/more.svg");
@@ -727,6 +740,17 @@
   mask-position-x: 0;
 }
 
+/* Translate the dropmarker to give illusion of increasing width of the recommendation  */
+#urlbar[cfr-recommendation-state] .urlbar-history-dropmarker {
+  transition: transform ease-in-out var(--cfr-animation-duration);
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker {
+  transform: translateX(calc(var(--cfr-label-width) * -1));
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker:-moz-locale-dir(rtl) {
+  transform: translateX(calc(var(--cfr-label-width)));
+}
+
 /* Shift the url-bar text fading to stop the recommendation overlapping */
 #urlbar[cfr-recommendation-state] #urlbar-input {
   /* A mask-image is usually only defined for the url bar when text overflows.
-- 
2.28.0

